File: algorithm/segment-tree/segment-tree.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> tree;

int init(vector<int> arr, int node_index, int left, int right) {
	if (left == right) {
		return tree[node_index] = arr[left];
	}

	int mid = (left + right) / 2;
	return tree[node_index] = 
		init(arr, node_index * 2, left, mid) +
		init(arr, node_index * 2 + 1, mid + 1, right);
}

void update(int node_index, int index, int left, int right, int diff) {
	if (!(left <= index && index <= right)) {
		return;
	}

	tree[node_index] += diff;

	if (left != right) {
		int mid = (left + right) / 2;
		update(node_index * 2, index, left, mid, diff);
		update(node_index * 2 + 1, index, mid + 1, right, diff);
	}
}

int query(int node_index, int a, int b, int left, int right) {
	if (b < left || right < a) {
		return 0;
	}

	if (a <= left && right <= b) {
		return tree[node_index];
	}

	int mid = (left + right) / 2;
	return query(node_index * 2, a, b, left, mid) +
		query(node_index * 2 + 1, a, b, mid + 1, right);
}

int main() {
	vector<int> arr = { 3, 5, 6, 7, 2, 9, 4, 5, 2, 8, 1, 5 };
	int left = 0;
	int right = arr.size() - 1;
	int root = 1;

	tree.reserve(4 * right);

	init(arr, root, left, right);
	cout << query(root, 0, 2, left, right) << '\n'; // 14 (3 + 5 + 6)

	update(root, 1, left, right, +6); // 3 -> 9
	cout << query(root, 0, 2, left, right) << '\n'; // 20 (9 + 5 + 6)

	update(root, 2, left, right, -2); // 5 -> 3
	cout << query(root, 0, 2, left, right) << '\n'; // 18 (9 + 3 + 6)

	cout << query(root, 3, 3, left, right) << '\n'; // 7

	return 0;
}

File: algorithm/segment-tree/segment-tree.ts
export default class SegmentTree {
	public tree: number[];
	private arr: number[];

	private root = 1;
	private left = 0;
	private right = 0;

	constructor(arr: number[]) {
		this.tree = new Array<number>(arr.length * 4);
		this.arr = arr;

		this.right = arr.length - 1;
		this.init(this.root, this.left, this.right);
	}

	public update(idx: number, newValue: number) {
		this._update(
			this.root,
			idx + 1, // root index is 1, not 0
			this.left,
			this.right,
			newValue - this.arr[idx]
		);
	}

	public query(from: number, to: number): number {
		return this._query(this.root, from, to, this.left, this.right);
	}

	private init(nodeIdx: number, left: number, right: number): number {
		if (left == right) {
			return (this.tree[nodeIdx] = this.arr[left]);
		}

		const mid = Math.floor((left + right) / 2);
		return (this.tree[nodeIdx] =
			this.init(nodeIdx * 2, left, mid) +
			this.init(nodeIdx * 2 + 1, mid + 1, right));
	}

	private _update(
		nodeIdx: number,
		idx: number,
		left: number,
		right: number,
		diff: number
	) {
		if (!(left <= idx && idx <= right)) return;

		this.tree[nodeIdx] += diff;

		if (left != right) {
			const mid = Math.floor((left + right) / 2);
			this._update(nodeIdx * 2, idx, left, mid, diff);
			this._update(nodeIdx * 2 + 1, idx, mid + 1, right, diff);
		}
	}

	private _query(
		nodeIdx: number,
		a: number,
		b: number,
		left: number,
		right: number
	): number {
		if (b < left || right < a) return 0;
		if (a <= left && right <= b) return this.tree[nodeIdx];

		const mid = Math.floor((left + right) / 2);
		return (
			this._query(nodeIdx * 2, a, b, left, mid) +
			this._query(nodeIdx * 2 + 1, a, b, mid + 1, right)
		);
	}
}

const segTree = new SegmentTree([3, 5, 6, 7, 2, 9, 4, 5, 2, 8, 1, 5]);
console.log(segTree.query(0, 2)); // 14

segTree.update(0, 9); // 3 -> 9
console.log(segTree.query(0, 2)); // 20

segTree.update(1, 3); // 5 -> 3
console.log(segTree.query(0, 2)); // 18

console.log(segTree.query(3, 3)); // 7
